src/extra/utils/ftd.pyx

```diff
@@ -0,0 +1,517 @@
+        interpolation points actually used to evaluate $x(t), i.e. the
+        finite boundaries to approximate the infinite sum above. By
+        default, up to 200 samples in each direction are used.
+
+    * `search_iterations` specifies the number of binary search steps
+        taken to find the optimal edge position from interpolated values
+        in between two samples. The maximal resolution in samples the
+        interpolation can therefore achieve is $2^-{\rm search_iterations}$.
+
+    When set, these parameters apply to all discriminator
+    implementations and all their use of sinc interpolation.
+
+    Args:
+        window (int, optional): Sample window used around the
+            interpolated point, unchanged if omitted.
+        search_iterations (int, optional): Number of iterations used in
+            binary search for closest function argument, unchanged if
+            omitted.
+
+    Returns
+        (window, search_iterations) Tuple of current values.
+    """
+
+    global sinc_window, sinc_search_iterations
+
+    if window is not None:
+        sinc_window = <int>window
+
+    if search_iterations is not None:
+        sinc_search_iterations = <int>search_iterations
+
+    return dict(window=sinc_window, search_iterations=sinc_search_iterations)
+
+
+cdef inline double _sinc(double x) nogil:
+    """Normalized sinc function."""
+
+    if x != 0.0:
+        return sin(x*M_PI) / (x*M_PI)
+    else:
+        return 1.0
+
+
+cdef double _sinc_interp(data_t[::contiguous] y_sampled, double x_interp) nogil:
+    cdef int k, \
+        sampling_start = max(<int>floor(x_interp) - sinc_window, 0), \
+        sampling_end = min(<int>ceil(x_interp) + sinc_window, y_sampled.shape[0])
+
+    cdef double y_interp = 0.0
+
+    for k in range(sampling_start, sampling_end):
+        y_interp += <double>y_sampled[k] * _sinc(k - x_interp)
+
+    return y_interp
+
+
+cdef data_t _cfd_sinc_interp(
+    data_array_t signal, int i, int j,
+    data_t delay, data_t fraction, data_t walk
+) nogil:
+    cdef int int_delay = <int>ceil(delay)
+    cdef bint is_integer_delay = int_delay == delay
+
+    cdef int k, \
+        sampling_start = max(i - sinc_window, 0), \
+        sampling_end = min(j + sinc_window, signal.shape[0])
+
+    # Interpolation is always done on double precision, as single
+    # precision is almost certain to cause rounding errors in the sum.
+    cdef double left_pos = <double>i, right_pos = <double>j, \
+        middle_pos = 0.0, middle_value
+
+    cdef data_t *interp_buf = NULL
+
+    if not is_integer_delay:
+        # For non-integer delays, the sinc interpolation includes
+        # the same interpolated values in its own summation for every
+        # iteration. Computing these values only once and keep them in
+        # a small buffer significantly increases performance.
+        interp_buf = <data_t*>malloc(sizeof(delay) * (2 * sinc_window + 1))
```

Comment:
Do we need to check for `malloc` returning NULL?

Reply:
I think so. Shows how long I have not used `malloc`.